File: loom-compiler/src/main/java/loom/compiler/token/Token.java
package loom.compiler.token;

import java.util.Objects;

public class Token implements Comparable<Token> {

	public final TokenType type;
	public final String value;
	public final TokenSpan span;

	public Token (TokenType type, String value, TokenSpan span) {
		this.type = type;
		this.value = value;
		this.span = span;
	}

	public boolean isOfType (TokenType expectedType) {
		return this.type == expectedType;
	}

	public int getLength () {
		return span.length();
	}

	public String getPositionString () {
		return span.getPositionString();
	}

	public TokenSpan getSpan () {
		return span;
	}

	@Override
	public String toString () {
		return String.format("TOKEN [%s, value='%s', %s]",
				type.toString().toUpperCase(), value, span.toString());
	}

	@Override
	public boolean equals (Object obj) {
		if (this == obj) return true;
		if (!(obj instanceof Token other)) return false;
		return type == other.type &&
				Objects.equals(value, other.value) &&
				Objects.equals(span, other.span);
	}

	@Override
	public int hashCode () {
		return Objects.hash(type, value, span);
	}

	@Override
	public int compareTo (Token other) {
		int result = Integer.compare(this.span.startLine(), other.span.startLine());
		if (result == 0) result = Integer.compare(this.span.startColumn(), other.span.startColumn());
		return result;
	}
}

File: loomc/lib/token.ml
(** Token representation for the Loom language lexer *)

type t = {
  token_type : Token_types.t;
  value : string;
  span : Token_span.t;
}

(** Create a token *)
let make token_type value span = { token_type; value; span }

(** Get token type *)
let token_type token = token.token_type

(** Get token value *)
let value token = token.value

(** Get token span *)
let span token = token.span

(** Check if token is of a specific type *)
let is_of_type token expected_type = token.token_type = expected_type

(** Get token length *)
let length token = Token_span.length token.span

(** Get position string for error reporting *)
let position_string token = Token_span.position_string token.span

(** Convert token to string representation *)
let to_string token =
  Printf.sprintf "TOKEN [%s, value='%s', %s]"
    (String.uppercase_ascii (Token_types.to_string token.token_type))
    token.value
    (Token_span.to_string token.span)

(** Compare tokens by position *)
let compare token1 token2 =
  let result = compare token1.span.start_line token2.span.start_line in
  if result = 0 then
    compare token1.span.start_column token2.span.start_column
  else
    result 